In [1]:
import pandas as pd
import pymongo, tqdm
from collections import defaultdict

client = pymongo.MongoClient('mongodb://localhost:27017')

# OLD DB (5.0.3)

In [86]:
import tqdm
def sorted_pair(s, o):
   return "_".join(sorted([s, o]))

db = client['DrugBanks']
collection = db['DB_503']
#cursor = collection.find({'$or':[{"s":{"$in":wanted_drugs_all}}, 
#                                 {"o":{"$in":wanted_drugs_all}}]})

cursor = collection.find()

print("Duplicate Pairs")
n_old = 0
found_old = []
for doc in tqdm.tqdm_notebook(cursor):
    n_old += 1
    cur = sorted_pair(doc['s'], doc['o'])
    found_old.append(cur)

print "Αll pairs: %d" % n_old
found_old = list(set(found_old))
print "Unique pairs: %d" %len(found_old)

unique_dbs_old = set()
for k in set(found_old):
    unique_dbs_old.add(k.split("_")[0])
    unique_dbs_old.add(k.split("_")[1])
print "Unique DB ids: %d" %len(unique_dbs_old)

Duplicate Pairs



Αll pairs: 581055
Unique pairs: 296326
Unique DB ids: 2551


# NEW DB

In [87]:
db = client['DrugBanks']
collection = db['DB_531']
# cursor = collection.find({'$or':[{"s":{"$in":wanted_drugs_all}}, 
#                                  {"o":{"$in":wanted_drugs_all}}]})
cursor = collection.find()
print("Duplicate Pairs")
n = 0
found = []
for doc in tqdm.tqdm_notebook(cursor):
    n += 1
    cur = sorted_pair(doc['s'], doc['o'])
    found.append(cur)

print "Αll pairs: %d" % n
found = list(set(found))
print "Unique pairs: %d" %len(found)

unique_dbs = set()
for k in set(found):
    unique_dbs.add(k.split("_")[0])
    unique_dbs.add(k.split("_")[1])
print "Unique DB ids: %d" %len(unique_dbs)

Duplicate Pairs



Αll pairs: 2641889
Unique pairs: 1392809
Unique DB ids: 3814


In [88]:
intersection = set(found).intersection(found_old)
new_not_in_old = set(found).difference(found_old)
old_not_in_new = set(found_old).difference(found)
# SANITY CHECK. EITHER IN 5.0.3 or not
assert len(new_not_in_old) + len(intersection) == len(found)
# SANITY CHECK. EITHER IN 5.1.3 or not (THIS DON'T KNOW WHY, BUT SHOULD BE 0!)
assert len(new_not_in_old) + len(intersection) == len(found)

print("DB 5.1.3: %d ALL | %d (%0.2f %%) UNIQUE DDIS. ALSO IN DB 5.0.3: %d (%0.2f %%) | NEW: %d (%0.2f %%)" % (n, len(found), 100*len(found)/float(n), len(intersection), 100*len(intersection)/float(len(found)), len(new_not_in_old), 100*len(new_not_in_old)/float(len(found))))

print("DB 5.0.3: %d ALL | %d (%0.2f %%) UNIQUE DDIS. ALSO IN DB 5.1.3: %d (%0.2f %%) | NEW: %d (%0.2f %%)" % (n_old, len(found_old), 100*len(found_old)/float(n_old), len(intersection), 100*len(intersection)/float(len(found_old)), len(old_not_in_new), 100*len(old_not_in_new)/float(len(found_old))))


DB 5.1.3: 2641889 ALL | 1392809 (52.72 %) UNIQUE DDIS. ALSO IN DB 5.0.3: 236174 (16.96 %) | NEW: 1156635 (83.04 %)
DB 5.0.3: 581055 ALL | 296326 (51.00 %) UNIQUE DDIS. ALSO IN DB 5.1.3: 236174 (79.70 %) | NEW: 60152 (20.30 %)


# SAVE CSVs

### First save:
- as train the unique old *296326*
- test all the unique new ones not in old *1156635*

In [93]:
new_.shape

(1156635, 3)

In [89]:
old_ = [pair.split('_') for pair in sorted(found_old)]
old_ = pd.DataFrame(old_, columns=['h','t'])
old_['r'] = 4
old_ = old_[['h', 'r', 't']]
old_.to_csv("./train_OLD_STR_ALL.txt", header=False, index=False)
assert old_.shape[0] == len(found_old)

In [90]:
new_ = [pair.split('_') for pair in sorted(new_not_in_old)]
new_ = pd.DataFrame(new_, columns=['h','t'])
new_['r'] = 4
new_ = new_[['h', 'r', 't']]
new_.to_csv("./test_CLEAN_STR_ALL.txt", header=False, index=False)
assert new_.shape[0] == len(new_not_in_old)

# Filter using the Drugs that we have description for


In [46]:
import pandas as pd
df = pd.read_csv('../DB_BioBERT_TorchKGE/DB_Bert_Emb.csv', encoding='utf8')
db_ids = sorted(df['DB_ID'].unique().tolist())
db_name_2_id = dict(zip(db_ids,range(len(db_ids))))
print("%d Unique Drugs!" % len(db_ids))

12112 Unique Drugs!


In [65]:
def check_in_biobert(row, db_ids=db_ids):
    if row['h'] in db_ids and row['t'] in db_ids:
        return 1
    else:
        return 0
    
old_['BIOBERT'] = old_.apply(check_in_biobert, axis=1)
print("OLD WITH BIOBERT DETAILS:")
print(old_['BIOBERT'].value_counts())
new_['BIOBERT'] = new_.apply(check_in_biobert, axis=1)
print("NEW WITH BIOBERT DETAILS:")
print(new_['BIOBERT'].value_counts())


/home/kostas/.local/lib/python2.7/site-packages/ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  import sys


OLD WITH BIOBERT DETAILS:
1    290722
0      5588
Name: BIOBERT, dtype: int64
NEW WITH BIOBERT DETAILS:
1    1004167
0     152468
Name: BIOBERT, dtype: int64


### WILL SAVE THOSE WITH BIOBERT EMBEDDINGS (BOTH DRUGS!)
- as train the unique old *290722* 
- test all the unique new ones not in old *1004167*

In [66]:
old_ = old_[old_["BIOBERT"]==1]
new_ = new_[new_["BIOBERT"]==1]
del old_['BIOBERT']
del new_['BIOBERT']
print("WILL SAVE %d UNIQUE OLD ONES FOR TRAIN" % (old_.shape[0]))
print("WILL SAVE %d UNIQUE NEW ONES FOR TEST"% (new_.shape[0]))
old_.to_csv("./train_OLD_STR_BIOBERT.txt", header=False, index=False)
new_.to_csv("./test_CLEAN_STR_BIOBERT.txt", header=False, index=False)

WILL SAVE 290722 UNIQUE OLD ONES FOR TRAIN
WILL SAVE 1004167 UNIQUE NEW ONES FOR TEST


# Save the BIOBERT ONES USING INTEGER MAPPING

In [73]:
import numpy as np
# First get all pairs (train+test) so as to find which db id's are actually used!
db_ids_with_biobert = np.concatenate([old_['h'].unique(),old_['t'].unique(),new_['h'].unique(),new_['t'].unique()])
df_new = df[df["DB_ID"].isin(db_ids_with_biobert)].reset_index(drop=True)
db_name_2_id_new = dict(zip(df_new.DB_ID.values,range(len(df_new.DB_ID.values))))
df_new["DB_ID"] = df_new.index.values


In [81]:
#old_['h'] = old_['h'].map(db_name_2_id_new)
#old_['t'] = old_['t'].map(db_name_2_id_new)
new_['h'] = new_['h'].map(db_name_2_id_new)
new_['t'] = new_['t'].map(db_name_2_id_new)

In [84]:
old_.to_csv("./train_OLD_INT_BIOBERT.txt", header=False, index=False)
new_.to_csv("./test_CLEAN_INT_BIOBERT.txt", header=False, index=False)
df_new.to_csv("./DB_Bert_Emb_INT.csv", index=False, header=True)

In [107]:
#id_2_db_new = dict(zip(db_name_2_id_new.values(), db_name_2_id_new.keys()))
#df_new["DB_ID"] = df_new['DB_ID'].map(id_2_db_new)
df_new.to_csv("./DB_Bert_Emb_STR.csv", index=False, header=True)

In [101]:
pd.DataFrame({'DB_ID':db_name_2_id_new.keys(), "INDEX":db_name_2_id_new.values()}).sort_values('DB_ID').to_csv("./DB_Bert_Emb_INT__MAPPING.csv", index=False)

In [106]:
df_new

0         1         2         3         4         5         6  \
0     0.202025 -0.006617 -0.099413  0.048500  0.060521 -0.073021 -0.090681   
1     0.177672  0.026007 -0.048232  0.018856  0.078177  0.000059 -0.337690   
2     0.242237  0.051406 -0.031407  0.186637  0.066632  0.044037 -0.095817   
3     0.226416  0.026122 -0.053552  0.251231  0.134691  0.050495 -0.148269   
4     0.257275 -0.033044 -0.031118  0.195232  0.170565 -0.095780 -0.081382   
5     0.089872 -0.104614 -0.017005  0.118278  0.163183 -0.079545 -0.081156   
6     0.030359  0.049868 -0.055352  0.058295  0.021681 -0.107721 -0.170107   
7     0.258122 -0.010933  0.012392  0.174634  0.166051 -0.054834 -0.148571   
8     0.262155  0.004658 -0.080868  0.239816  0.321597 -0.123344  0.008519   
9    -0.008348  0.101089 -0.040840  0.216455  0.040123  0.119392 -0.093857   
10    0.227782  0.023204 -0.134949  0.278564  0.267014 -0.092425 -0.028497   
11    0.093290 -0.055498 -0.007089  0.138134  0.138491 -0.087377 -0.137931   
12    0.267202 -0.035367 -0.014268  0.227959  0.186943 -0.123303 -0.092012   
13    0.017521  0.044465  0.129911  0.020036  0.124795  0.024742 -0.114001   
14    0.054093 -0.016334 -0.019077  0.020072  0.255939  0.006307 -0.077289   
15    0.260311  0.006016 -0.077675  0.242775  0.324749 -0.127825  0.012286   
16    0.360440  0.003908  0.094224  0.135690  0.035062  0.016181 -0.220587   
17    0.109307 -0.030422 -0.097475  0.135617  0.060250 -0.049899 -0.268757   
18    0.061381  0.051031 -0.077537  0.041240  0.044710 -0.096930 -0.177771   
19    0.060101 -0.141942 -0.059566 -0.063811 -0.053714 -0.120051 -0.070622   
20    0.066802 -0.090746 -0.072274  0.202641  0.001624  0.023289 -0.152888   
21   -0.007509 -0.091119 -0.096133  0.277252  0.056154 -0.079019 -0.083552   
22    0.216094  0.024181  0.096357  0.054192 -0.089694 -0.054780 -0.206116   
23    0.013039 -0.008013 -0.113393  0.014451 -0.033504 -0.087944 -0.178109   
24    0.104064 -0.059583 -0.145407  0.141507  0.064965  0.033490 -0.042832   
25    0.255918 -0.089850 -0.094315  0.193061  0.184082 -0.105772 -0.075587   
26    0.209576  0.064942 -0.154443  0.236067  0.299258 -0.008045 -0.062795   
27    0.294117 -0.025825 -0.019896  0.291086  0.164942 -0.064182 -0.120736   
28    0.267238  0.115723 -0.060110  0.211331  0.108443 -0.009139 -0.161534   
29    0.122481 -0.012198  0.061073  0.143448 -0.130966 -0.001296 -0.182079   
...        ...       ...       ...       ...       ...       ...       ...   
3649  0.240105 -0.243725 -0.031099  0.160957  0.099914 -0.199394  0.239319   
3650  0.206333 -0.234239 -0.056064  0.188663  0.136652 -0.190827  0.259927   
3651  0.047498 -0.011358  0.035957  0.302181  0.086886 -0.037021 -0.069925   
3652 -0.088854 -0.239162  0.063202  0.430630  0.117073 -0.128405 -0.113518   
3653  0.216840 -0.357858  0.110138  0.246034  0.223447 -0.335794  0.255071   
3654  0.243045 -0.360706  0.034855  0.215249  0.157562 -0.298133  0.251841   
3655  0.236017 -0.339457  0.056437  0.265844  0.176514 -0.319493  0.226570   
3656  0.196267 -0.364210  0.048894  0.282791  0.228269 -0.325612  0.229509   
3657  0.049734 -0.198366 -0.048279  0.161488  0.236212 -0.178519  0.041248   
3658 -0.055121 -0.142932 -0.194469  0.146961  0.087231 -0.103806 -0.148434   
3659  0.224567 -0.285636  0.109514  0.264723  0.169367 -0.276045  0.202686   
3660  0.075023 -0.166977 -0.005551  0.179157  0.073182 -0.112625  0.018543   
3661  0.351030  0.008219  0.144115  0.097681  0.140037  0.066006 -0.250854   
3662  0.019607 -0.279212 -0.177176  0.179841  0.141391 -0.065032 -0.054013   
3663  0.165818 -0.288263  0.025394  0.242070  0.187887 -0.259788  0.332988   
3664  0.125706  0.053955 -0.160698  0.133392  0.225559 -0.104228 -0.242414   
3665  0.126405 -0.040554  0.091628  0.025451  0.195871  0.029603 -0.053231   
3666  0.183980 -0.337376  0.007502  0.202899  0.169276 -0.253334  0.318380   
3667  0.199465 -0.277384  0.039785  0.204232  0.139953 -0.262794  0.334410   
3668  0.103030 -0.149711 -0.081474 